presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,472 @@
+    }
+
+    @Override
+    public Optional<JdbcExpression> implementAggregation(ConnectorSession session, AggregateFunction aggregate,
+            Map<String, ColumnHandle> assignments)
+    {
+        return aggregateFunctionRewriter.rewrite(session, aggregate, assignments);
+    }
+
+    private static Optional<JdbcTypeHandle> toTypeHandle(DecimalType decimalType)
+    {
+        return Optional.of(new JdbcTypeHandle(Types.NUMERIC, Optional.of("decimal"), decimalType.getPrecision(),
+                decimalType.getScale(), Optional.empty(), Optional.empty()));
+    }
+
+    @Override
+    public void abortReadConnection(Connection connection)
+            throws SQLException
+    {
+        connection.abort(directExecutor());
+    }
+
+    private String quoted(@Nullable String schema, String table)
+    {
+        StringBuilder sb = new StringBuilder();
+        if (!isNullOrEmpty(schema)) {
+            sb.append(quoted(schema)).append(".");
+        }
+        sb.append(quoted(table));
+        return sb.toString();
+    }
+
+    @Override
+    public void finishInsertTable(ConnectorSession session, JdbcOutputTableHandle handle)
+    {
+        String temporaryTable = quoted(handle.getSchemaName(), handle.getTemporaryTableName());
+        String targetTable = quoted(handle.getSchemaName(), handle.getTableName());
+        String columnNames = handle.getColumnNames().stream().map(this::quoted).collect(joining(", "));
+        String insertSql = format("INSERT INTO %s (%s) SELECT %s FROM %s", targetTable, columnNames, columnNames,
+                temporaryTable);
+        String cleanupSql = "DROP TABLE " + temporaryTable;
+
+        try (Connection connection = getConnection(session, handle)) {
+            execute(connection, insertSql);
+        }
+        catch (SQLException e) {
+            throw new PrestoException(JDBC_ERROR, e);
+        }
+
+        try (Connection connection = getConnection(session, handle)) {
+            execute(connection, cleanupSql);
+        }
+        catch (SQLException e) {
+            log.warn(e, "Failed to cleanup temporary table: %s", temporaryTable);
+        }
+    }
+
+    @Override
+    public PreparedStatement buildSql(ConnectorSession session, Connection connection, JdbcSplit split,
+            JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        // All remote table become default.schema.table
+        RemoteTableName remoteTableName = new RemoteTableName(Optional.empty(),
+                Optional.ofNullable(table.getSchemaName()), table.getTableName());
+        return new QueryBuilder(this).buildSql(session, connection, remoteTableName, table.getGroupingSets(), columns,
+                table.getConstraint(), split.getAdditionalPredicate(), tryApplyLimit(table.getLimit()));
+    }
+
+    @Override
+    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, Connection connection,
```

Comment:
Please add `TestClickHouseTypeMapping`. You can copy `TestPostgreSqlTypeMapping` as the base of the test.

Reply:
OK, I will add the testing